src/wasm-type.h

```diff
@@ -251,17 +261,82 @@ struct ResultType {
   std::string toString() const;
 };
 
+class HeapType {
+  // HeapTypes are canonicalized and interned exactly like Types.
+  uintptr_t id;
+
+public:
+  enum BasicHeapType : uint32_t {
+    func,
+    ext,
+    exn,
+    any,
+    eq,
+    i31,
+  };
+  static constexpr BasicHeapType _last_basic_type = i31;
+
+  // BasicHeapType can be implicitly upgraded to HeapType
+  constexpr HeapType(BasicHeapType id) : id(id) {}
+
+  // But converting raw uint64_t is more dangerous, so make it explicit
+  explicit HeapType(uint64_t id) : id(id) {}
+
+  HeapType(Signature signature);
+  HeapType(const Struct& struct_);
+  HeapType(Struct&& struct_);
+  HeapType(Array array);
+
+  constexpr bool isBasic() const { return id <= _last_basic_type; }
+  constexpr bool isCompound() const { return id > _last_basic_type; }
+  bool isFunction() const;
+  bool isSignature() const;
+  bool isStruct() const;
+  bool isArray() const;
+
+  Signature getSignature() const;
+  const Struct& getStruct() const;
+  Array getArray() const;
```

Comment:
Is the reason Arrays and Structs differ in this way that Arrays are "small" and Structs are possibly of unbounded size?

Reply:
Yes, that's what I was thinking 👍 Arrays contain only one `Field` each, which in turn contains only three items taking up two machine words. That seems small enough to pass around by value.

Reply:
That does make sense, but it does make the API asymmetrical, which seems a little surprising. Or do you think for users of the API the difference wouldn't be noticeable?

A comment might be good either way.

Reply:
Yeah, the asymmetry is a little unfortunate, but I think it will be ok. What would you think of deleting `operator=` to prevent folks from accidentally doing `auto s = getStruct()` and making a copy of the struct? Would that be worth it or too inconvenient?

Will add comments.

Reply:
That seems good, assuming I understand correctly and it would just force people to add the `&` in `auto& struct_ = type.getStruct();` - ? If so sgtm. But regardless, that could also be separate.